codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +70,1044 @@
+                    }
+                    return true;
+                }
+            });
+        } catch (Http2Exception e) {
+            // TODO: log me ?
+            ctx.close();
+        } finally {
+            // We need to ensure we release the goAwayFrame.
+            goAwayFrame.release();
         }
     }
 
     /**
      * Notifies any child streams of the read completion.
      */
     @Override
-    public void channelReadComplete(ChannelHandlerContext ctx) {
-        for (int i = 0; i < channelsToFireChildReadComplete.size(); i++) {
-            Http2StreamChannel childChannel = channelsToFireChildReadComplete.get(i);
-            // Clear early in case fireChildReadComplete() causes it to need to be re-processed
-            childChannel.inStreamsToFireChildReadComplete = false;
-            childChannel.fireChildReadComplete();
-        }
-        channelsToFireChildReadComplete.clear();
+    public final void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+        onChannelReadComplete(ctx);
+        channelReadComplete0(ctx);
     }
 
-    ChannelFuture createStreamChannel(Channel parentChannel, EventLoopGroup group, ChannelHandler handler,
-                                              Map<ChannelOption<?>, Object> options,
-                                              Map<AttributeKey<?>, Object> attrs,
-                                              int streamId) {
-        final Http2StreamChannel channel = new Http2StreamChannel(parentChannel);
-        if (isStreamIdValid(streamId)) {
-            assert !isOutboundStream(server, streamId);
-            assert ctx.channel().eventLoop().inEventLoop();
-            channel.streamId(streamId);
+    final void onChannelReadComplete(ChannelHandlerContext ctx)  {
+        // If we have many child channel we can optimize for the case when multiple call flush() in
+        // channelReadComplete(...) callbacks and only do it once as otherwise we will end-up with multiple
+        // write calls on the socket which is expensive.
+        try {
+            DefaultHttp2StreamChannel current = head;
+            while (current != null) {
+                DefaultHttp2StreamChannel childChannel = current;
+                if (childChannel.fireChannelReadPending) {
+                    // Clear early in case fireChildReadComplete() causes it to need to be re-processed
+                    childChannel.fireChannelReadPending = false;
+                    if (childChannel.fireChildReadComplete()) {
+                        flushNeeded = true;
+                    }
+                }
+                childChannel.next = null;
+                current = current.next;
+            }
+        } finally {
+            // We always flush as this is what Http2ConnectionHandler does for now.
+            // TODO: I think this is not really necessary and we should be able to optimize this in the future.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)